ts/client/services/events/genesisBatch.ts

```diff
@@ -0,0 +1,124 @@
+import { ContractEventSyncer } from "./contractEventSyncer";
+
+const genesisBatchID = 0;
+
+/**
+ * Syncs genesis newBatch events from the rollup contract
+ */
+export class GenesisBatchEventSyncer extends ContractEventSyncer {
+    private readonly genesisStateRoot: string;
+    private readonly rollup: Rollup;
+    private readonly syncpoint: SyncedPoint;
+    private readonly batchStorage: BatchStorage;
+
+    constructor(api: CoreAPI) {
+        super(
+            api.contracts.rollup,
+            api.contracts.rollup.filters.NewBatch(null, null, null)
+        );
+        this.eventListener = this.genesisBatchListener;
+
+        this.genesisStateRoot = api.getGenesisRoot();
+        this.rollup = api.contracts.rollup;
+        this.syncpoint = api.syncpoint;
+        this.batchStorage = api.l2Storage.batches;
+    }
+
+    public async initialSync(
+        startBlock: number,
+        endBlock: number
+    ): Promise<void> {
+        const events = await this.getEvents(startBlock, endBlock);
+        const genesisBatchEvents = events.filter(this.isGenesisBatchEvent);
+        if (!genesisBatchEvents.length) {
+            return;
+        }
+
+        if (genesisBatchEvents.length > 1) {
+            throw new TooManyGenesisBatchesFound(genesisBatchEvents.length);
+        }
+        await this.handleGenesisBatch(genesisBatchEvents[0]);
+    }
+
+    private isGenesisBatchEvent(e: Event): boolean {
+        return (e.args?.batchType as Usage) === Usage.Genesis;
+    }
+
+    private validateIsFirstBatch(batchID: number) {
+        if (batchID !== genesisBatchID) {
+            throw new NotFirstBatch();
+        }
+    }
+
+    private async validateNotAlreadySynced() {
+        const existingGenesisBatch = await this.batchStorage.getByID(
+            genesisBatchID
+        );
+        if (existingGenesisBatch) {
+            const existingL1Txn = await this.batchStorage.getL1Transaction(
+                existingGenesisBatch.commitmentRoot
+            );
+            throw new GenesisBatchAlreadySynced(existingL1Txn?.hash);
+        }
+    }
+
+    private async handleGenesisBatch(event: Event) {
+        const batchIDBN = event.args?.batchID as BigNumber;
+        const batchID = batchIDBN.toNumber();
+
+        this.validateIsFirstBatch(batchID);
+        await this.validateNotAlreadySynced();
+
+        // Get L1 state
+        const l1Txn = await event.getTransaction();
+        const l1GenesisBatch = await this.rollup.getBatch(batchIDBN, {
+            blockTag: l1Txn.blockNumber
+        });
+
+        // Create L2 state
+        const genCommit = new GenesisCommitment(this.genesisStateRoot);
+        const genesisBatch = new ConcreteBatch([genCommit]);
+
+        // Verify match
+        if (l1GenesisBatch.commitmentRoot !== genesisBatch.commitmentRoot) {
+            throw new GenesisBatchCommitmentRootMismatch(
+                l1GenesisBatch.commitmentRoot,
+                genesisBatch.commitmentRoot
+            );
```

Comment:
This and the validations/errors above are all the edge cases I could think of to verify the genesis batch, let me know if you know of other cases/things we should verify.